File: services/crypto-core/ratchet.go
package cryptocore

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

const (
	hkdfInfoRatchet       = "SecuMSG-DR"
	maxSkippedMessageKeys = 64
)

// Encrypt derives the next sending message key, returns the ciphertext and the
// message header that must accompany the ciphertext.
func Encrypt(session *SessionState, plaintext []byte) ([]byte, *MessageHeader, error) {
	if session == nil {
		return nil, nil, errors.New("cryptocore: nil session")
	}
	if isZeroKey(session.SendChain.Key) {
		if err := RotateRatchetOnSend(session, nil); err != nil {
			return nil, nil, err
		}
	}
	newCK, mk := kdfChain(session.SendChain.Key)
	n := session.SendChain.Index
	session.SendChain.Key = newCK
	session.SendChain.Index++

	key, nonce, err := deriveCipherParams(mk)
	if err != nil {
		return nil, nil, err
	}
	aead, err := chacha20poly1305.New(key[:])
	if err != nil {
		return nil, nil, err
	}
	header := &MessageHeader{DHPublic: session.RatchetPublic, PN: session.PN, N: n, Nonce: nonce}
	ad := header.associatedData()
	ciphertext := aead.Seal(nil, nonce[:], plaintext, ad)
	return ciphertext, header, nil
}

// Decrypt attempts to open the ciphertext using the provided header, handling
// skipped message keys as necessary.
func Decrypt(session *SessionState, ciphertext []byte, header *MessageHeader) ([]byte, error) {
	if session == nil {
		return nil, errors.New("cryptocore: nil session")
	}
	if header == nil {
		return nil, errors.New("cryptocore: nil header")
	}
	if mk, ok := session.consumeSkipped(header); ok {
		key, nonce, err := deriveCipherParams(mk)
		if err != nil {
			return nil, err
		}
		aead, err := chacha20poly1305.New(key[:])
		if err != nil {
			return nil, err
		}
		plaintext, err := aead.Open(nil, nonce[:], ciphertext, header.associatedData())
		if err != nil {
			return nil, ErrDecryptionFailed
		}
		return plaintext, nil
	}
	if err := RotateRatchetOnRecv(session, header); err != nil {
		return nil, err
	}
	if header.N < session.RecvChain.Index {
		return nil, ErrDuplicateMessage
	}
	for session.RecvChain.Index < header.N {
		newCK, mk := kdfChain(session.RecvChain.Key)
		session.storeSkippedKey(session.RemoteRatchet, session.RecvChain.Index, mk)
		session.RecvChain.Key = newCK
		session.RecvChain.Index++
	}
	newCK, mk := kdfChain(session.RecvChain.Key)
	session.RecvChain.Key = newCK
	session.RecvChain.Index++
	key, nonce, err := deriveCipherParams(mk)
	if err != nil {
		return nil, err
	}
	aead, err := chacha20poly1305.New(key[:])
	if err != nil {
		return nil, err
	}
	plaintext, err := aead.Open(nil, nonce[:], ciphertext, header.associatedData())
	if err != nil {
		return nil, ErrDecryptionFailed
	}
	return plaintext, nil
}

// RotateRatchetOnSend advances the local sending ratchet and prepares a fresh
// sending chain based on a newly generated DH key pair.
func RotateRatchetOnSend(session *SessionState, _ *MessageHeader) error {
	if session == nil {
		return errors.New("cryptocore: nil session")
	}
	if isZeroKey(session.RemoteRatchet) {
		return ErrInvalidRemoteKey
	}
	kp, err := generateX25519KeyPair()
	if err != nil {
		return err
	}
	dh, err := curve25519.X25519(kp.Private[:], session.RemoteRatchet[:])
	if err != nil {
		return err
	}
	root, send, err := kdfRoot(session.RootKey[:], dh)
	if err != nil {
		return err
	}
	session.RootKey = root
	session.PN = session.SendChain.Index
	session.SendChain = chainState{Key: send, Index: 0}
	session.RatchetPrivate = kp.Private
	session.RatchetPublic = kp.Public
	return nil
}

// RotateRatchetOnRecv processes a header received from the remote party and
// updates the receiving chain if a new remote ratchet key is presented.
func RotateRatchetOnRecv(session *SessionState, header *MessageHeader) error {
	if session == nil {
		return errors.New("cryptocore: nil session")
	}
	if header == nil {
		return errors.New("cryptocore: nil header")
	}
	if header.DHPublic == session.RemoteRatchet {
		return nil
	}
	dh, err := curve25519.X25519(session.RatchetPrivate[:], header.DHPublic[:])
	if err != nil {
		return err
	}
	root, recv, err := kdfRoot(session.RootKey[:], dh)
	if err != nil {
		return err
	}
	session.RootKey = root
	session.RemoteRatchet = header.DHPublic
	session.RecvChain = chainState{Key: recv, Index: 0}
	session.SendChain.Key = [32]byte{}
	session.SendChain.Index = 0
	session.PN = header.PN
	return nil
}

func kdfRoot(root, dh []byte) ([32]byte, [32]byte, error) {
	hk := hkdf.New(sha256.New, dh, root, []byte(hkdfInfoRatchet))
	var newRoot, chain [32]byte
	if _, err := io.ReadFull(hk, newRoot[:]); err != nil {
		return [32]byte{}, [32]byte{}, err
	}
	if _, err := io.ReadFull(hk, chain[:]); err != nil {
		return [32]byte{}, [32]byte{}, err
	}
	return newRoot, chain, nil
}

func kdfChain(chain [32]byte) ([32]byte, [32]byte) {
	mk := hmacSHA256(chain[:], []byte{0x01})
	out := hmacSHA256(chain[:], []byte{0x02})
	var next, msg [32]byte
	copy(next[:], mk)
	copy(msg[:], out)
	return next, msg
}

func hmacSHA256(key, data []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}

func deriveCipherParams(mk [32]byte) ([32]byte, [12]byte, error) {
	hk := hkdf.New(sha256.New, mk[:], nil, []byte("SecuMSG-AEAD"))
	var key [32]byte
	var nonce [12]byte
	if _, err := io.ReadFull(hk, key[:]); err != nil {
		return [32]byte{}, [12]byte{}, err
	}
	if _, err := io.ReadFull(hk, nonce[:]); err != nil {
		return [32]byte{}, [12]byte{}, err
	}
	return key, nonce, nil
}

func (h *MessageHeader) associatedData() []byte {
	buf := make([]byte, 32+4+4)
	copy(buf, h.DHPublic[:])
	binary.BigEndian.PutUint32(buf[32:], h.PN)
	binary.BigEndian.PutUint32(buf[36:], h.N)
	return buf
}

func isZeroKey(k [32]byte) bool {
	var zero [32]byte
	return k == zero
}

func (s *SessionState) storeSkippedKey(pub [32]byte, index uint32, key [32]byte) {
	if s.skipped == nil {
		s.skipped = make(map[string][32]byte)
	}
	if len(s.skipped) >= maxSkippedMessageKeys {
		for k := range s.skipped {
			delete(s.skipped, k)
			break
		}
	}
	name := skippedKey(pub, index)
	s.skipped[name] = key
}

func (s *SessionState) consumeSkipped(header *MessageHeader) ([32]byte, bool) {
	if s.skipped == nil {
		return [32]byte{}, false
	}
	name := skippedKey(header.DHPublic, header.N)
	if val, ok := s.skipped[name]; ok {
		delete(s.skipped, name)
		return val, true
	}
	return [32]byte{}, false
}

func skippedKey(pub [32]byte, index uint32) string {
	buf := make([]byte, 32+4)
	copy(buf, pub[:])
	binary.BigEndian.PutUint32(buf[32:], index)
	return string(buf)
}

File: frontend/src/crypto-core/ratchet.ts
import { ChaCha20Poly1305 } from "@stablelib/chacha20poly1305";
import { curve25519 } from "@noble/curves/curve25519";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { hmac } from "@noble/hashes/hmac";
import { generateX25519KeyPair } from "./core";
import {
  ErrDecryptionFailed,
  ErrDuplicateMessage,
  ErrInvalidRemoteKey,
} from "./errors";
import type { MessageHeader, SessionState } from "./types";
import { copyBytes, equalsBytes, zeroBytes, utf8 } from "./utils";

const hkdfInfoRatchet = "SecuMSG-DR";
const hkdfInfoAEAD = "SecuMSG-AEAD";
const maxSkippedMessageKeys = 64;

export function Encrypt(
  session: SessionState,
  plaintext: Uint8Array,
): { ciphertext: Uint8Array; header: MessageHeader } {
  if (!session) {
    throw new Error("cryptocore: nil session");
  }
  if (isZeroKey(session.SendChain.Key)) {
    RotateRatchetOnSend(session, undefined);
  }
  const { chainKey: newCK, messageKey: mk } = kdfChain(session.SendChain.Key);
  const n = session.SendChain.Index;
  session.SendChain.Key = newCK;
  session.SendChain.Index += 1;

  const { key, nonce } = deriveCipherParams(mk);
  const aead = new ChaCha20Poly1305(key);
  const header = new MessageHeader(copyBytes(session.RatchetPublic), session.PN, n, nonce);
  const ad = header.associatedData();
  const ciphertext = aead.seal(nonce, plaintext, ad);
  return { ciphertext, header };
}

export function Decrypt(session: SessionState, ciphertext: Uint8Array, header: MessageHeader): Uint8Array {
  if (!session) {
    throw new Error("cryptocore: nil session");
  }
  if (!header) {
    throw new Error("cryptocore: nil header");
  }
  const skipped = consumeSkipped(session, header);
  if (skipped) {
    const { key, nonce } = deriveCipherParams(skipped);
    const aead = new ChaCha20Poly1305(key);
    try {
      return aead.open(nonce, ciphertext, header.associatedData());
    } catch {
      throw ErrDecryptionFailed;
    }
  }
  RotateRatchetOnRecv(session, header);
  if (header.N < session.RecvChain.Index) {
    throw ErrDuplicateMessage;
  }
  while (session.RecvChain.Index < header.N) {
    const { chainKey: newCK, messageKey } = kdfChain(session.RecvChain.Key);
    storeSkippedKey(session, session.RemoteRatchet, session.RecvChain.Index, messageKey);
    session.RecvChain.Key = newCK;
    session.RecvChain.Index += 1;
  }
  const { chainKey: newCK, messageKey: mk } = kdfChain(session.RecvChain.Key);
  session.RecvChain.Key = newCK;
  session.RecvChain.Index += 1;
  const { key, nonce } = deriveCipherParams(mk);
  const aead = new ChaCha20Poly1305(key);
  try {
    return aead.open(nonce, ciphertext, header.associatedData());
  } catch {
    throw ErrDecryptionFailed;
  }
}

export function RotateRatchetOnSend(session: SessionState, _header?: MessageHeader): void {
  if (!session) {
    throw new Error("cryptocore: nil session");
  }
  if (isZeroKey(session.RemoteRatchet)) {
    throw ErrInvalidRemoteKey;
  }
  const kp = generateX25519KeyPair();
  const dh = curve25519.scalarMult(kp.Private, session.RemoteRatchet);
  const { root, chain } = kdfRoot(session.RootKey, dh);
  session.RootKey = root;
  session.PN = session.SendChain.Index;
  session.SendChain = { Key: chain, Index: 0 };
  session.RatchetPrivate = copyBytes(kp.Private);
  session.RatchetPublic = copyBytes(kp.Public);
}

export function RotateRatchetOnRecv(session: SessionState, header: MessageHeader): void {
  if (!session) {
    throw new Error("cryptocore: nil session");
  }
  if (!header) {
    throw new Error("cryptocore: nil header");
  }
  if (equalsBytes(header.DHPublic, session.RemoteRatchet)) {
    return;
  }
  const dh = curve25519.scalarMult(session.RatchetPrivate, header.DHPublic);
  const { root, chain } = kdfRoot(session.RootKey, dh);
  session.RootKey = root;
  session.RemoteRatchet = copyBytes(header.DHPublic);
  session.RecvChain = { Key: chain, Index: 0 };
  session.SendChain = { Key: zeroBytes(32), Index: 0 };
  session.PN = header.PN;
}

function kdfRoot(root: Uint8Array, dh: Uint8Array): { root: Uint8Array; chain: Uint8Array } {
  const okm = hkdf(sha256, dh, root, utf8(hkdfInfoRatchet), 64);
  return {
    root: okm.slice(0, 32),
    chain: okm.slice(32, 64),
  };
}

function kdfChain(chain: Uint8Array): { chainKey: Uint8Array; messageKey: Uint8Array } {
  const mk = hmac(sha256, chain, new Uint8Array([0x01]));
  const out = hmac(sha256, chain, new Uint8Array([0x02]));
  return {
    chainKey: mk.slice(0, 32),
    messageKey: out.slice(0, 32),
  };
}

function deriveCipherParams(messageKey: Uint8Array): { key: Uint8Array; nonce: Uint8Array } {
  const okm = hkdf(sha256, messageKey, new Uint8Array(), utf8(hkdfInfoAEAD), 44);
  return {
    key: okm.slice(0, 32),
    nonce: okm.slice(32, 44),
  };
}

function isZeroKey(key: Uint8Array): boolean {
  return key.every((b) => b === 0);
}

function storeSkippedKey(
  session: SessionState,
  pub: Uint8Array,
  index: number,
  key: Uint8Array,
): void {
  if (!session.skipped) {
    session.skipped = new Map();
  }
  if (session.skipped.size >= maxSkippedMessageKeys) {
    const firstKey = session.skipped.keys().next().value;
    if (firstKey) {
      session.skipped.delete(firstKey);
    }
  }
  session.skipped.set(skippedKey(pub, index), copyBytes(key));
}

function consumeSkipped(session: SessionState, header: MessageHeader): Uint8Array | undefined {
  if (!session.skipped || session.skipped.size === 0) {
    return undefined;
  }
  const name = skippedKey(header.DHPublic, header.N);
  const val = session.skipped.get(name);
  if (val) {
    session.skipped.delete(name);
    return copyBytes(val);
  }
  return undefined;
}

function skippedKey(pub: Uint8Array, index: number): string {
  const buf = new Uint8Array(36);
  buf.set(pub, 0);
  const view = new DataView(buf.buffer, buf.byteOffset + 32, 4);
  view.setUint32(0, index, false);
  let out = "";
  for (let i = 0; i < buf.length; i += 1) {
    out += String.fromCharCode(buf[i]);
  }
  return out;
}
